### ISSUE TITLE
[libc++] False positive mutex '...' is still held at the end of function

### ISSUE THREAD

**Issue by PiJoules** [llvm/llvm-project#156760]
```
#include <mutex>

std::mutex m0, m1;

void func() {
  std::unique_lock ul1(m0, std::defer_lock);
  std::unique_lock ul2(m1, std::defer_lock);
  std::lock(ul1, ul2);
}
```

```
$ ~/misc/clang-cipd-latest/bin/clang++ -c -Wall -Wthread-safety /tmp/test.cc
/tmp/test.cc:9:1: warning: mutex 'ul1' is still held at the end of function [-Wthread-safety-analysis]
    9 | }
      | ^
/tmp/test.cc:8:3: note: mutex acquired here
    8 |   std::lock(ul1, ul2);
      |   ^
/tmp/test.cc:9:1: warning: mutex 'ul2' is still held at the end of function [-Wthread-safety-analysis]
    9 | }
      | ^
/tmp/test.cc:8:3: note: mutex acquired here
    8 |   std::lock(ul1, ul2);
      |   ^
2 warnings generated.
```

This comes up with https://github.com/llvm/llvm-project/pull/154078/. Both mutexes should be unlocked by the `unique_lock` dtors at the end of the function. Also `-Wthread-safety-analysis` is incorrectly saying `ul1` and `ul2` are mutexes.

**Comment by aaronpuchert** [comment]
> Both mutexes should be unlocked by the unique_lock dtors at the end of the function.

The [current code](https://github.com/llvm/llvm-project/blob/main/libcxx/include/__mutex/unique_lock.h) does not have any annotation for that:
```c++
  _LIBCPP_HIDE_FROM_ABI ~unique_lock() {
    if (__owns_)
      __m_->unlock();
  }
```
It also seems that `unique_lock` isn't annotated as `scoped_capability`. This is already an issue.

Another problem appears with #154078: `std::lock` is not only meant for what the analysis calls lockable types, but also scoped lockable types such as `unique_lock`. But we don't support that, not even on the attribute level. You should see a warning like "'acquire_capability' attribute requires arguments whose type is annotated with 'capability' attribute; type here is 'unique_lock' [-Wthread-safety-attributes]".

What we need here is to add support for that.
